### ISSUE TITLE
Cryptography Support

### ISSUE THREAD

**Issue by psytron** [pyodide/pyodide#761]
Pyodide is genius. It will certainly influence and broaden the world impact of all these data science tools and Python in so many new contexts. 

Any plan on adding 'cryptography' to the list of supported packages?

It seems many cool modules depend on it for fundamental encryption and security. All applications that would be awesome in the browser with Pyodide. 

e.g.
```console
>>> micropip.install('ccxt')
Couldn't find a pure Python 3 wheel for 'cryptography'
>>> micropip.install('alias')
Couldn't find a pure Python 3 wheel for 'cryptography'
>>> micropip.install('keycache')
Couldn't find a pure Python 3 wheel for 'cryptography'
```

**Comment by rth** [comment]
The issue with adding the cryptography package is that it uses CFFI for C-extensions and we currently don't support CFFI (https://github.com/iodide-project/pyodide/issues/681#issuecomment-639413545).

**Comment by RoryShively** [comment]
Is this a limitation of pyodide or would we be able to add this in with a PR?

**Comment by mdtanrikulu** [comment]
Same applies for [`pynacl`](https://github.com/pyca/pynacl/) package. It took some time to figure out that it uses CFFI (for libsodium), and pyodide is not able to load dynamic libraries for that reason. It would be amazing to be able to use such packages.

**Comment by Kaiyiccc** [comment]
Same for google/tink btw:
```
ValueError: Couldn't find a pure Python 3 wheel for 'tink'
```

**Comment by hoodmane** [comment]
Note that we do support cffi since #728.

**Comment by rth** [comment]
I meant the https://pypi.org/project/cffi package not the C FFI in general. I don't know what would be the practical difference in making it work as compared to ctypes, but we should try adding it as a first step then.

**Comment by hoodmane** [comment]
I see. Well it looks like libffi support was a blocker. Looking at the cffi github there's a lot of stuff that I don't understand at all. It might well be quite a complex task still.

**Comment by hoodmane** [comment]
Okay well cffi only asks for the libffi library, so there is a chance it might just work if we try to build it.

**Comment by hoodmane** [comment]
I am trying to build it and so far I am running into the problem that it is trying to link with `-lffi` but `libffi` can only be linked into the main module. Somehow need to modify config to filter out that flag or convince it not to issue it.

**Comment by hoodmane** [comment]
We're presumably going to start having trouble when cffi tries to compile things, since we don't have a C compiler.

**Comment by hoodmane** [comment]
Okay well even with the cffi blocker cleared, there is a bigger problem which is that `cryptography` has a rust/pyo3 module. Building rust modules seems to be a pretty daunting task.

**Comment by rth** [comment]
It's by no mean a long term solution but we could try building cryptography 3.3.2 (from Feb 7, 2021) which was before they added the rust/pyo3  dependency.

**Comment by hoodmane** [comment]
That is probably a good idea. Another concern about rust is that it generates very large build artifacts: a Rust Python module that defines a single API that returns a constant ends compiles to a 3.4Mb .so, though obviously the size is something we can concern ourselves with we figure out how to actually build stuff...

**Comment by hoodmane** [comment]
Perhaps I'll open a draft PR with some of my partial progress toward building rust modules. I might be getting close, a lot of problems come up though.

**Comment by hoodmane** [comment]
Next issue with cryptography is lots of `six`.

**Comment by rth** [comment]
> some of my partial progress toward building rust modules.

Even if it works, and personally I like playing with Rust, I would be a bit hesitant to add it to the default build. People have already enough trouble building Pyodide as is, without adding the Rust dependency. Or at least we would need to revisit the idea that all included packages are built by default.

> a Rust Python module that defines a single API that returns a constant compiles to a 3.4Mb .so

Hmm, interesting. I was under mistaken impression that Rust would make smaller wasm binaries given all the work being done in that area. But even https://rustpython.github.io/demo/ is still 10MB compressed, which is larger than us.

**Comment by hoodmane** [comment]
>  at least we would need to revisit the idea that all included packages are built by default.

Yeah

> I was under mistaken impression that Rust would make smaller wasm binaries

I think they compress it very well, it just comes with TONS of extra stuff in it by default. For instance, all of the fancy panic messages. Their `print!` macro generates a whole pile of support code. Generics rapidly increase code size. In C there just aren't any constructs that generate enormous amounts of code behind the scenes. I think if you strip out all of this stuff, it might make a smaller binary but it is not clear how to do it. It also might not be fair because I made no effort at all to shrink this. 

In any case, Rust seems much harder to port than C.

**Comment by hoodmane** [comment]
Okay so next issue: we probably need to port openSSL for this.

**Comment by planctron** [comment]
There should be a "pure" python implementation. Did you look for such alternative?
Why? It's even possible to implement PGP in Javascript.
See   https://openpgpjs.org/

**Comment by hoodmane** [comment]
@planctron Yeah that would be a reasonable approach. One would have to implement [custom backend interfaces](https://cryptography.io/en/latest/hazmat/backends/interfaces/) with pgpjs. I most likely won't do this myself but if someone is interested in contributing that would be welcome.

**Comment by hoodmane** [comment]
@planctron In the original post, the rational given for wanting the cryptography package is that a large number of other packages are dependent on it (ccxt, alias, and keycache are given as examples).

**Comment by psytron** [comment]
Ok, good points. Brain storming; Would it be possible to wrap those async methods in promises and then use 'await' inside the wrapper methods?  ( To make them sync.)  Or compile the original cryptography library within the webassembly dependency but map the missing strong entropy to use subtle.crypto instead of whatever is causing the build to not work in webassembly?

**Comment by rth** [comment]
I think the best way forward is still to build the original cryptography library version 3.4.8 (2021-08-24) before rust was added as a build requirement. To do that one would need to build openssl as a dynamic library https://github.com/pyodide/pyodide/issues/2124. As far as I can tell that's the only remaining blocker.

**Comment by hoodmane** [comment]
I think I can get cryptography to work. They actually introduced the rust dependency in v3.4.0 so we will need to use v3.3.2 until we can resolve https://github.com/rust-lang/rust/issues/93015 or hack our way around it.

**Comment by hoodmane** [comment]
Well I successfully built and imported cryptography v3.3.2, but there are some dynamic linking errors that need to be debugged.

**Comment by luigiw** [comment]
Hello, sounds like at least cryptography v3.3.2 should work. Any suggestion on why I got a complaining can't find a pure python package for cryptography>=2.1.4?

```
File "/lib/python3.9/site-packages/micropip/micropip.py", line 253, in add_wheel
    await self.add_requirement(recurs_req, ctx, transaction)
  File "/lib/python3.9/site-packages/micropip/micropip.py", line 241, in add_requirement
    wheel, ver = self.find_wheel(metadata, req)
  File "/lib/python3.9/site-packages/micropip/micropip.py", line 269, in find_wheel
    raise ValueError(f"Couldn't find a pure Python 3 wheel for '{req}'")
ValueError: Couldn't find a pure Python 3 wheel for 'cryptography>=2.1.4'
```

**Comment by ryanking13** [comment]
It seems like you are using an older version of Pyodide. Cryptography (v3.3.2) is included from Pyodide 0.20.0.
